src/engine/Match_tainting_mode.ml

```diff
@@ -648,6 +734,15 @@ let check_rules ~match_hook
        RP.rule_profiling RP.match_result) (rules : R.taint_rule list)
     (xconf : Match_env.xconfig) (xtarget : Xtarget.t) :
     RP.rule_profiling RP.match_result list =
+  (* We create a "formula cache" here, before dealing with individual rules, to
+     permit sharing of matches for sources, sanitizers, propagators, and sinks
+     between rules.
+
+     In particular, this expects to see big gains due to shared propagators,
+     in Semgrep Pro. There may be some benefit in OSS, but it's low-probability.
```

Comment:
I wouldn't say it's low probability, have you tested it?

Reply:
i haven't, but i find the likelihood of a source, sink, or sanitizer being the exact same between different rules being a relatively chance event (outside of hardcoded propagators). note that they need to be _exactly_ the same, even down to things like metavariable naming

Reply:
Asked SR but my intuition is that there should be, because rules for the same framework should use mostly the same sources and sanitizers. And I guess SR just copy-pastes them, although perhaps not always. Almost as soon as SR started using taint mode we started discussing how to share sources across taint rules.